core/server/api/themes.js

```diff
@@ -26,23 +23,41 @@ var debug = require('debug')('ghost:api:themes'),
  */
 themes = {
     browse: function browse() {
-        debug('browsing');
-        var result = themeList.toAPI(themeList.getAll(), settingsCache.get('activeTheme'));
-        debug('got result');
-        return Promise.resolve({themes: result});
+        return Promise.resolve(themeUtils.toJSON());
     },
 
     activate: function activate(options) {
         var themeName = options.name,
             newSettings = [{
                 key: 'activeTheme',
                 value: themeName
-            }];
+            }],
+            loadedTheme,
+            checkedTheme;
+
+        return apiUtils
+            .handlePermissions('themes', 'activate')(options)
+            .then(function activateTheme() {
+                loadedTheme = themeList.get(themeName);
+
+                if (!loadedTheme) {
+                    return Promise.reject(new errors.ValidationError({
+                        message: i18n.t('notices.data.validation.index.themeCannotBeActivated', {themeName: themeName}),
+                        context: 'activeTheme'
+                    }));
+                }
 
-        return settings.edit({settings: newSettings}, options).then(function () {
-            var result = themeList.toAPI(themeList.getAll(), themeName);
-            return Promise.resolve({themes: result});
-        });
+                return themeUtils.validate.check(loadedTheme);
+            })
+            .then(function haveValidTheme(_checkedTheme) {
+                checkedTheme = _checkedTheme;
+                // We use the model, not the API here, as we don't want to trigger permissions
+                return settingsModel.edit(newSettings, options);
+            })
+            .then(function hasEditedSetting() {
+                // @TODO actually do things to activate the theme, other than just the setting?
```

Comment:
What else need's to be done here except of setting the active theme in the model?